arch/mips/core/thread.c

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2017 Imagination Technologies Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <arch/cpu.h>
+#include <kernel_structs.h>
+#include <wait_q.h>
+#include <string.h>
+#include <soc.h>
+
+void _thread_entry_wrapper(k_thread_entry_t thread,
+			   void *arg1,
+			   void *arg2,
+			   void *arg3);
+
+void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
+		 size_t stack_size, k_thread_entry_t entry,
+		 void *arg1, void *arg2, void *arg3,
+		 int prio, unsigned int options)
+{
+	_ASSERT_VALID_PRIO(prio, entry);
+
+	char *stack_memory = K_THREAD_STACK_BUFFER(stack);
+	char *ctx = (char *)STACK_ROUND_DOWN(stack + stack_size -
+		sizeof(struct gpctx));
+
+	thread->callee_saved.sp = (u32_t)ctx;
+
+	struct gpctx stack_init = {
+		.r = {
+			[3] = (reg_t)entry,
+			[4] = (reg_t)arg1,
+			[5] = (reg_t)arg2,
+			[6] = (reg_t)arg3,
+			[27] = mips32_get_gp(),
+			[28] = (reg_t)ctx
+		},
+		/* core timer and sw0 interrupts default to on EXL is
+		 * also set asstatus will be written to this value
+		 * inside an ISR
+		 */
+		.status = (reg_t)0x20008103,
+		.cause = mips_getcr() & ~(CR_IP_MASK),
+		.epc = (reg_t)_thread_entry_wrapper
```

Comment:
As mentioned earlier, since _thread_entry_wrapper doesn't seem to do much at all, can we just directly put _thread_entry here instead of _thread_entry_wrapper?